File: leetcode/98-validate-binary-search-tree.py
# https://leetcode.com/problems/validate-binary-search-tree/ (top-interview-questions)

from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # run inorder traverse and check if the values are sorted
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        values_inorder = list()
        stack = list()
        node = root

        while stack or node:
            while node:
                stack.append(node)
                node = node.left
            node = stack.pop()
            values_inorder.append(node.val)
            node = node.right
        
        for i in range(1, len(values_inorder)):
            if values_inorder[i-1] >= values_inorder[i]:
                return False
        
        return True
    


    def is_valid_bst(self, root: Optional[TreeNode], low=(-2)**31-1, high=(2**31)) -> bool:
        if not root: 
            return True
        if root.val <= low or root.val >= high:
            return False

        return (self.is_valid_bst(root.left, low, root.val) and self.is_valid_bst(root.right, root.val, high))

File: leetcode/98-validate-binary-search-tree.js
// https://leetcode.com/problems/validate-binary-search-tree/ (top-interview-questions)


// Definition for a binary tree node.
function TreeNode(val, left, right) {
  this.val = (val===undefined ? 0 : val)
  this.left = (left===undefined ? null : left)
  this.right = (right===undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {

  function validation(node, max, min) {
    if(node == null)
        return true;
    if(node.val <= min || node.val >= max)
        return false;
    
    return validation(node.left, node.val, min) && validation(node.right, max, node.val);
  }
  
  return validation(root, Infinity, -Infinity);
};
